### ISSUE TITLE
ECNRSigner can create unverifiable signature

### ISSUE THREAD

**Issue by tonywasher** [bcgit/bc-java#562]
It is possible to create a signature via ECNRSigner that will fail validation. 

The problem can be seen for some signatures when the digest size is identical to the order of the curve, for example the TigerDigest and the brainpoolP192t1 curve. The trigger for the failure is when the calculated digest is transformed into a BigInteger and results in a value that is larger than the N value for the curve.

Since the validation process calculates a result modulo N before comparing to the digest, it is not possible for the validation to succeed in this scenario. 

The comments in the ECNRSigner say

> Generally, the order of the curve should be at 
     least as long as the hash of the message of interest, and with 
     ECNR, it *must* be at least as long.  But just in case the signer
     applied mod(n) to the longer digest, this implementation will
     apply mod(n) during verification.

It would appear that  you need to apply mod(n) to the digest as well.

Code to reproduce the problem follows 

`  public static void main(final String[] pArgs) {

        /* Create the generator */
        ECKeyPairGenerator myGenerator = new ECKeyPairGenerator();
        SecureRandom myRandom = new SecureRandom();
        String myCurve = "brainpoolP192t1";

        /* Lookup the parameters */
        final X9ECParameters x9 = ECNamedCurveTable.getByName(myCurve);

        /* Initialise the generator */
        final ASN1ObjectIdentifier myOid = ECUtil.getNamedCurveOid(myCurve);
        ECNamedDomainParameters myDomain = new ECNamedDomainParameters(myOid, x9.getCurve(), x9.getG(), x9.getN(), x9.getH(), x9.getSeed());
        ECKeyGenerationParameters myParams = new ECKeyGenerationParameters(myDomain, myRandom);
        myGenerator.init(myParams);

        /* Create the key Pair */
        AsymmetricCipherKeyPair myPair = myGenerator.generateKeyPair();

        /* Create the digest and the output buffer */
        Digest myDigest = new TigerDigest();
        byte[] myArtifact = new byte[myDigest.getDigestSize()];
        final byte[] myMessage = "Hello there. How is life treating you?".getBytes();
        myDigest.update(myMessage, 0, myMessage.length);
        myDigest.doFinal(myArtifact, 0);

        /* Create signer */
        ECNRSigner mySigner = new ECNRSigner();
        mySigner.init(true, myPair.getPrivate());
        BigInteger[] mySignature = mySigner.generateSignature(myArtifact);

        /* verify signature */
        mySigner.init(false, myPair.getPublic());
        if (!mySigner.verifySignature(myArtifact, mySignature[0], mySignature[1])) {
            System.out.println("ECNR failed");
        }
    }
`

**Comment by bcgit** [comment]
Thanks for this. So the JavaDoc was incorrect, but now is. I've also fixed the input bounds check as well so that out of range input is rejected. ECNR is supposed to be a message recovery scheme so a msg larger than the order the curve should not be used, I think the confusion has come about as it's also used as a general signature scheme. The changes will mean the JavaDoc is now correct, but the signer will not produce signatures that may result in other implementations mistaking the message if recovery is being used as well (it's still possible to produce such signatures, but the message has to be explicitly truncated by the developer). I've added a method to the low level API to support recovery based applications as well. Up in the current betas are 163b10 or later.

PS. I replied to an email on feedback-crypto recently. Did you receive the reply?
